rotkehlchen/globaldb/upgrades/v3_v4.py

```diff
@@ -256,17 +258,69 @@ def _add_optimism_contracts(cursor: 'DBCursor', eth_scan_abi_id: int, multicall_
+        assets_ids: list[str],
+        root_dir: Path,
+) -> None:
+    """
+    Copy missing assets from the packaged globaldb to the globaldb
+    Since the upgrade happens when the connection to the globaldb hasn't been created yet we need
+    to manually copy the assets information as we don't have the GlobalDBHandler object.
+
+    We need to commit the changes here because otherwise the packaged globaldb is busy and we
+    can't detach it. At this point all the information in the upgrade should be fine.
+
+    TODO: This approach assumes that the schema in the packaged globaldb and the schema in v3 are
+    the same. When this doesn't hold anymore we should keep a copy of the old schema or ask users
+    to use an intermediate version to upgrade.
+    """
+    packaged_db_path = root_dir / 'data' / 'global.db'
+    identifiers_quotes = ",".join("?" * len(assets_ids))
+    cursor.execute(f'ATTACH DATABASE "{packaged_db_path}" AS packaged_db;')
+    cursor.execute(f'INSERT INTO assets SELECT * FROM packaged_db.assets WHERE identifier IN ({identifiers_quotes});', assets_ids)  # noqa: E501
+    cursor.execute(f'INSERT INTO evm_tokens SELECT * FROM packaged_db.evm_tokens WHERE identifier IN ({identifiers_quotes});', assets_ids)  # noqa: E501
+    cursor.execute(f'INSERT INTO common_asset_details SELECT * FROM packaged_db.common_asset_details WHERE identifier IN ({identifiers_quotes});', assets_ids)  # noqa: E501
+    cursor.execute('COMMIT;')
+    cursor.execute('DETACH DATABASE packaged_db;')
+
+
 def _add_assets_collections(cursor: 'DBCursor') -> None:
     """
     Insert in the global db the assets collection properties for the assets we know and
     create the collections.
+
+    If any of the assets that needs to go in the collections is missing we copy it from the
+    packaged globaldb.
     """
     log.debug('Enter _add_assets_collections')
     root_dir = Path(__file__).resolve().parent.parent.parent
+    # regex used to identify assets that will be added to the collections
+    asset_regex = re.compile(r'eip155[a-zA-F0-9:\/]+')
     with open(root_dir / 'data' / 'assets_collections.sql') as f:
         raw_sql_sentences = f.read()
         per_table_sentences = raw_sql_sentences.split('\n\n')
         for sql_sentences in per_table_sentences:
+            # check if we are adding the assets
+            if 'multiasset_mappings' in sql_sentences:
+                # in this case we need to ensure that the assets exist locally and
+                # if not copy them from the packaged db
+                all_assets = asset_regex.findall(sql_sentences)
+                assets_to_add = []
+                for asset_identifier in all_assets:
+                    cursor.execute(
+                        'SELECT COUNT(*) FROM assets where identifier=?',
+                        (asset_identifier,),
+                    )
+                    if cursor.fetchone()[0] == 0:
+                        assets_to_add.append(asset_identifier)
```

Comment:
silly goose :smile_cat:!
You don't need to iterate in python and make as many DB queries as there is assets. You can just do 1 query.

The below is untested pseudocode but I guess you get the idea


```suggestion
                cursor.execute('SELECT identifier from assets WHERE identifier IN (?, ?, ? .... )', all_assets)
                existing_ids = [x[0] for x in cursor]
                assets_to_add = set(all_assets) - set(existing_ids)
```

Reply:
indeed much better